### DISCUSSION TITLE
Meaning of `Function.assign(scalar)`

### DISCUSSION THREAD

**Discussion by jrmaddison** [firedrakeproject/firedrake#3245] [Firedrake support] [answered]
I have previously read `u.assign(1.0)` as 'set the `Function` `u` equal to the function equal to one'. This would only make sense for function spaces for which the function equal to one is a member -- e.g. this wouldn't make sense for a zero trace space, or for non scalar-valued spaces.

However I see e.g.

```
from firedrake import *

mesh = UnitSquareMesh(2, 2)
space = FunctionSpace(mesh, "RT", 1)
u = Function(space, name="u")

u.assign(1.0)
assert (u.dat.data_ro == 1.0).all()
```

runs without error, suggesting that `u.assign(1.0)` sets the _degrees of freedom_ to one -- a different and basis dependent operation.

`u.assign(Constant(1.0))` has the same behaviour.

Is this the intended behaviour, a bug, or am I using the `assign` method in an unexpected/invalid way?

**Comment by jrmaddison** [comment]
For comparison with legacy FEniCS
```
from fenics import *

mesh = UnitSquareMesh(2, 2)
space = FunctionSpace(mesh, "RT", 1)
u = Function(space, name="u")

u.assign(Constant((1.0, 1.0)))
print(f"{assemble(u[0] * dx)=}")
```
displays (approx) one. Using `u.assign(Constant(1.0))` leads to an error.

**Comment by colinjcotter** [comment]
It’s really a question of whether it is necessary to anticipate every mistake the user makes and catch it with an error.

**Reply by jrmaddison** [reply]
I guess that's the third option.

**Comment by wence-** [comment, marked as answer]
`.assign` does dof-wise assignment of linear combinations of things in "compatible" spaces. So this is working as intended. In almost all cases you want to use `.interpolate` if you actually care about function spaces and the meaning of the value (if that is too slow, please open a bug)

**Reply by dham** [reply]
I guess there is a case for saying that `.assign` should not take constant values (possibly except zero, and if the space is `R`)

**Reply by jrmaddison** [reply]
A more relevant example is

```
from firedrake import *

mesh = UnitSquareMesh(2, 2)
e = EnrichedElement(FiniteElement("Lagrange", mesh.ufl_cell(), 1),
                    FiniteElement("Bubble", mesh.ufl_cell(), 3))
space = FunctionSpace(mesh, e)

u = Function(space, name="u").assign(Constant(1.0))

print(f"{assemble(u * dx)=}")
```

with the result `assemble(u * dx)=1.4499999999999995`.

**Reply by wence-** [reply]
This is a _long_-standing consequence of `EnrichedElement` (it doesn't reorthonormalise the basis). If you want that, use `NodalEnrichedElement`

**Reply by jrmaddison** [reply]
I mean it's a continuous real scalar-valued case where dual basis evaluation is not point evaluation. dof assignment and function assignment both make sense here. It's not wrong, just not what I expected.

**Reply by wence-** [reply]
No, it's worse than that, `interpolate` into an enriched space like that will _also_ give you 1.5

**Reply by wence-** [reply]
Because the `EnrichedElement` implementation only gives you a prime basis for which $\phi_i^*[\phi_j] = \delta_{ij}$ if the sum of the individual bases is direct.
